src/main/kotlin/org/jitsi/nlj/rtcp/KeyframeRequester.kt

```diff
@@ -49,54 +52,31 @@ class KeyframeRequester : TransformerNode("Keyframe Requester") {
 
     override fun transform(packetInfo: PacketInfo): PacketInfo? {
         val packet = packetInfo.packet
-        val pliOrFir = when {
-            packet is CompoundRtcpPacket -> {
-                packet.packets.first { it is RtcpFbPliPacket || it is RtcpFbFirPacket }
+        val pliOrFir = when (packet) {
+            is CompoundRtcpPacket -> {
+                packet.packets.first { it is RtcpFbPliPacket || it is RtcpFbFirPacket } as? RtcpFbPacket
             }
-            packet is RtcpFbFirPacket -> packet
-            packet is RtcpFbPliPacket -> packet
+            is RtcpFbFirPacket -> packet
+            is RtcpFbPliPacket -> packet
             else -> null
-        } ?: return null
-
-        var forward = true
-        when {
-            pliOrFir is RtcpFbPliPacket -> {
-                when {
-                    hasPliSupport -> {
-                        // Forward PLI
-                        forward = canSendKeyframeRequest(pliOrFir.mediaSenderSsrc, System.currentTimeMillis())
-                    }
-                    hasFirSupport -> {
-                        // Translate to FIR
-                        requestKeyframe(pliOrFir.mediaSenderSsrc)
-                        forward = false
-                    }
-                    else -> {
-                        // Can't do anything
-                        forward = false
-                    }
-                }
-            }
-            pliOrFir is RtcpFbFirPacket -> {
-                when {
-                    hasPliSupport -> {
-                        // Translate to PLI
-                        requestKeyframe(pliOrFir.mediaSenderSsrc)
-                        forward = false
-                    }
-                    hasFirSupport -> {
-                        // Forward FIR
-                        forward = canSendKeyframeRequest(pliOrFir.mediaSenderSsrc, System.currentTimeMillis())
-                        if (forward) {
-                            pliOrFir.seqNum = firCommandSequenceNumber.incrementAndGet()
-                        }
-                    }
-                    else -> {
-                        // Can't do anything
-                        forward = false
-                    }
-                }
-            }
+        } ?: return packetInfo
+
+        val now = System.currentTimeMillis()
+        val canSend = canSendKeyframeRequest(pliOrFir.mediaSourceSsrc, now)
+        var forward = when (pliOrFir) {
+            is RtcpFbPliPacket -> hasPliSupport && canSend
+            // When both are supported, we favor generating a PLI rather than forwarding a FIR
+            is RtcpFbFirPacket -> hasFirSupport && !hasPliSupport && canSend
```

Comment:
will this get us what we want?  if it's an FIR packet, but PLI is supported we won't forward this FIR and won't to convert to a PLI?  or maybe i'm misreading

Reply:
ok, i think i see below this is handled by the code on line 79.

Reply:
If it's an FIR packet and we support PLI this line will set `forward` to `false`. Subsequently, if `canSend` is true, we will generate our own request (see logic below), which favors PLI.